onnxruntime/core/providers/cuda/math/softmax_impl.cu

```diff
@@ -39,67 +39,59 @@ Status dispatch_warpwise_softmax_forward(cudaStream_t stream, output_t* dst, con
 
     // This value must match the WARP_SIZE constexpr value computed inside softmax_warp_forward.
     int warp_size = (next_power_of_two < GPU_WARP_SIZE_HOST) ? next_power_of_two : GPU_WARP_SIZE_HOST;
-
-    // This value must match the WARP_BATCH constexpr value computed inside softmax_warp_forward.
+    int threads_per_block, shared_memory_size;
+    // there are 2 options to save one row of the input matrix: register or shared memory
+    // when the number of elements is small, we use register; otherwise, we use shared memory;
     int batches_per_warp = (next_power_of_two <= 128) ? 2 : 1;
-
-    // use 128 threads per block to maximimize gpu utilization
-    constexpr int threads_per_block = 128;
-
+    if (log2_elements <= 10){
+      // This value must match the WARP_BATCH constexpr value computed inside softmax_warp_forward.
+      // use 128 threads per block to maximimize gpu utilization
+      threads_per_block = 128;
+      shared_memory_size = 0;
```

Comment:
having shared_memory_size , means we don't set contraints on shared memory usage, right? If so, shall we make it clear in the comment also?

Reply:
"shared_memory_size" here means dynamic shared memory allocated when launching the kernel, the kernel itself can still allocate shared memory with static shap

Reply:
OK, my question is, why we need have the value for launching the kernel?

Reply:
what happens if we don't set it? fail?

Reply:
Here is the conclusion we talked offline: shared memory size can be allocated in the kernel code as a fixed size array, in that case, we don't need pass share memory size when launch the kernel. 
Another use case is, we allocate the share memory for each block during kernel launch, then in the kernel, we just get t he shared memory pointer and use the value with an appropriate index. This is the case we have in the PR.